### PR TITLE
Raster to NumPy Array Support

### PR DESCRIPTION
**Pull request by geordgez**
### Short description
Added option to output data as NumPy array in RasterFileIO in `gaia/geo/geo_inputs.py`. 

### Example Usage
Default functionality matches the NumPy array calls in the [`docs/examples/gaia_processes.ipynb`](https://github.com/OpenDataAnalytics/gaia/blob/master/docs/examples/gaia_processes.ipynb) notebook (example below):
~~~~
globaltemp = RasterFileIO(uri='../../tests/data/globalairtemp.tif')
temparray = np.array(globaltemp.read().GetRasterBand(1).ReadAsArray())
~~~~
With the new function the second line can be rewritten:
~~~~
temparray = globaltemp.read(as_numpy_array=True)
~~~~

### Parameter defaults & descriptions
~~~~
read(self, as_numpy_array=False, as_single_band=True, new_nodata=None, epsg=None)
~~~~

`as_single_band` parameter supports output of 3D NumPy array for multidimensional raster datasets (default is a 2D slice of the first raster band).
- If `as_single_band=False`, we can get the equivalent layer as above by indexing into the first layer of the 3D NumPy array:
~~~~
temparray = globaltemp.read(as_numpy_array=True, as_single_band=False)[0]
~~~~

`new_nodata` parameter supports customization of NoData values:
- If `new_nodata=None` (default value), attempts to get the NoData values in each band in the output NumPy array. If no NoData value exists, sets NoData value to 0.

### CODE REVIEW

**Comment by mbertrand** [comment]
Thanks @geordgez, looks good to me.  I've added @aashish24 as a reviewer for final approval.

**Comment by geordgez** [comment]
@aashish24: 
@mbertrand and I just had a discussion about how to handle NoData values within raster band layers. How should we fill out NoData in the NumPy array output, i.e., should we use `NaN`, `0`, or some other value?

**Comment by aashish24** [comment]
> NoData value

I would go with this order. Fill with original NoData value, if none found or none given from the user, then use NaN as the value. I would avoid 0 as 0 could be a valid value.

**Comment by aashish24** [comment]
@geordgez @mbertrand see this issue https://github.com/OpenDataAnalytics/gaia/issues/60

**Comment by geordgez** [comment]
Thanks for the info @aashish24! I just checked to make sure that the most recent commit (`54535f1`) includes the new parameters `old_nodata` and `new_nodata` for the `read` function with the functionality below in each situation (**8 unique cases** are covered). 

Is this consistent with the desired NoData functionality?

Situations:
- `(old_nodata is None) and (srcband.GetNoDataValue() is None)` or
`(new_nodata is None)` (**5 cases**)
  - no cells in the returned array `srcband_ar = srcband.ReadAsArray()` are replaced. Original NoData values remain unchanged.

- `(old_nodata is not None) and (new_nodata is not None)` (**2 cases**)
  - all cells that match `old_nodata` in the returned array `srcband.ReadAsArray()` are replaced with `new_nodata`

- `(old_nodata is None) and (srcband.GetNoDataValue() is not None) and (new_nodata is not None)` (**1 case**)
  - all cells that match the NoData value obtained from `srcband.GetNoDataValue()` in the returned array `srcband.ReadAsArray()` are replaced with `new_nodata`

Full function signature:
~~~~
read(
  self, 
  as_numpy_array=False, 
  as_single_band=True, 
  old_nodata=None, 
  new_nodata=None, 
  epsg=None
)
~~~~

**Comment by aashish24** [comment]
@geordgez thanks for putting it together very well. I do not quite understand this: 

(old_nodata is not None) and (new_nodata is not None) (2 cases)

How can we have two possibilities?

**Comment by aashish24** [comment]
@geordgez also any idea on how we can define a structure behind conversions?

**Comment by geordgez** [comment]
@aashish24 Sorry I should clarify, 8 cases are for the combinations of `old_nodata`, `new_nodata`, and `srcband.GetNoDataValue()` taking on values of `None` or `not None`. 

So the combination of `(old_nodata is not None)` and `(new_nodata is not None)` includes both possibilities that `(srcband.GetNoDataValue() is None)` and `(srcband.GetNoDataValue() is not None)`.

In terms of defining a structure behind conversions, do you mean for going both ways between raster Tiff and NumPy?

**Comment by aashish24** [comment]
> In terms of defining a structure behind conversions, do you mean for going both ways between raster Tiff and NumPy?

Sorry what I meant is if we should define a base class that defines an API for conversions? Since currently we have a module level function but in the future we may need conversion to more types (pandas df for example)

File: gaia/geo/geo_inputs.py [hunk 1]
```diff
@@ -215,10 +216,22 @@ class RasterFileIO(FileIO):
...
         Read data from a raster dataset
 
+        :param as_numpy_array: Output data as numpy
+        (default is False i.e. raster osgeo.gdal.Dataset)
+        :param as_single_band: Output data as 2D array of its first band
+        (default is True). If False, returns full 3D array.
+        :param old_nodata: Explicitly identify existing NoData values
+        (default None). If None, attempts to get existing NoData values stored
+        in the raster band.
+        :param new_nodata: Replace NoData values in each band with new_nodata
```
Comment (geordgez) [inline gaia/geo/geo_inputs.py]: Yeah I wasn't quite sure about the best way to logically to describe it. A sketch of my thoughts below:

- A user can try to get the original NoData value stored in raster band `b` by passing `old_nodata=None`.
- However, for a given raster band `b`, `b.GetNoDataValue()` doesn't always return a NoData value. Sometimes raster bands don't explicitly identify a NoData value, in which case `b.GetNoDataValue()` returns `None`.
- If the user knows in advance that a given value `d` is the NoData value within the band, then the user can help the `raster_to_numpy_ar` function by explicitly identifying that value. When a user does explicitly identify `d` as the NoData value then it overrides `b.GetNoDataValue()`, even if `b.GetNoDataValue()` exists (isn't `None`).

Does that make a bit more sense? Let me know if you have any thoughts on rephrasing the description in the docs.

File: gaia/geo/geo_inputs.py [hunk 2]
```diff
@@ -487,3 +505,45 @@ def reproject(dataset, epsg):
     elif dataclass == 'Dataset':
         repro = gdal_reproject(dataset, '', epsg=epsg)
     return repro
+
+
+def raster_to_numpy_ar(raster_data, as_single_band=True,
```
Comment (geordgez) [inline gaia/geo/geo_inputs.py]: Sounds good, I'll make the naming changes and push.

**Comment by geordgez** [comment]
> Sorry what I meant is if we should define a base class that defines an API for conversions? Since currently we have a module level function but in the future we may need conversion to more types (pandas df for example)

Understood--I think having a conversion class may be a good idea depending on the types of possible conversions within the program. On my end, I need to familiarize myself with the file formats and conversions that exist (or that we may want in the future). 

On the one hand, I think it will be helpful to consolidate all the conversion. On the other hand, I want to avoid a ambiguity where a user wants to make a standard conversion (e.g., from NumPy to Pandas) and is unsure whether to use the converter API or to use the standard Pandas function calls.

File: gaia/geo/geo_inputs.py (hunk 1 above)
Comment (aashish24) [inline gaia/geo/geo_inputs.py]: makes sense thanks @geordgez

**Comment by aashish24** [comment]
> On the one hand, I think it will be helpful to consolidate all the conversions. On the other hand, I want to avoid an ambiguity where a user wants to make a standard conversion (e.g., from NumPy to Pandas) and is unsure whether to use the converter API or to use the standard Pandas function calls.

sure. Did you get a chance to think more about it? I am thinking having a converted API would be nice since I am expecting that we will need that quite a bit in the future.

**Comment by geordgez** [comment]
@aashish24 Having given some thought, I think the converter API would be a good idea although I'd want to verify some aspects with you and the team. My initial questions/thoughts:
- Which formats would the API cover? Would it mainly be the raster formats found in gaia.formats?
- Based on what I've seen so far and with the functionality added by @andrenguyen-bah and @chuehlien, NumPy seems to be a "universal" intermediate format for conversion of typical image files since it can support all of the output image types. Would the API be built around intermediate stages with NumPy?
- Sorry if this is a dumb question: are there any instances where we would be converting between Raster, Feature, or Vector classes? Or are conversions always between formats within each class?

**Comment by geordgez** [comment]
See #92 and #60

**Comment by aashish24** [comment]
@geordgez will reply later today.

**Comment by aashish24** [comment]
@geordgez thinking some more on this, lets get merged this one and then we can make another pass on it.

**Comment by aashish24** [comment]
> Which formats would the API cover? Would it mainly be the raster formats found in gaia.formats?

for now raster, in the future, we should also cover vector (for example geotiff to vector format).

**Comment by aashish24** [comment]
> Based on what I've seen so far and with the functionality added by @andrenguyen-bah and @chuehlien, NumPy seems to be a "universal" intermediate format for conversion of typical image files since it can support all of the output image types. Would the API be built around intermediate stages with NumPy?

I think that should be fine. I think it is easy to convert from NumPy to other types such as pandas DF.

**Comment by aashish24** [comment]
> Sorry if this is a dumb question: are there any instances where we would be converting between Raster, Feature, or Vector classes? Or are conversions always between formats within each class?

should be able to go between types. See here: https://docs.qgis.org/2.6/en/docs/training_manual/complete_analysis/raster_to_vector.html

**Comment by aashish24** [comment]
@geordgez thoughts?

**Comment by aashish24** [comment]
Also, can you squash commits please? 

Thanks,

**Comment by geordgez** [comment]
@aashish24 Sorry for the delay, I got dragged away and likely won't be available for a few weeks--I definitely think going between types should be a capability of the API. Let met know if you need more squashes to commits or amendments to commit comments.

My main concern going forward is how we'd be expanding the API for objects in memory vs objects on disk and the relationship between the two object types, especially with how they interact with NumPy. Converting to raster formats through NumPy should be easy--correct me if I'm wrong--but I think we need to be careful with the process for converting to vector formats.

**Pull request merged**